File: leoz-node/src/main/java/org/deku/leoz/node/auth/Authorizer.java
package org.deku.leoz.node.auth;

import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;
import org.deku.leoz.messaging.MessagingContext;
import org.deku.leoz.messaging.activemq.ActiveMQContext;
import org.deku.leoz.node.LocalStorage;
import org.deku.leoz.node.messaging.auth.IdentityPublisher;
import org.deku.leoz.node.messaging.auth.v1.AuthorizationMessage;
import sx.Disposable;
import sx.Dispose;
import sx.jms.embedded.Broker;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/**
 * Created by masc on 01.07.15.
 */
public class Authorizer implements Disposable {
    Log mLog = LogFactory.getLog(this.getClass());

    /** Messaging context */
    private MessagingContext mMessagingContext;
    /** Executor service for authorization task */
    private ExecutorService mExecutorService;
    /** Authorization task */
    private Runnable mAuthorizationTask;

    private Broker.EventListener mBrokerEventListener = new Broker.EventListener() {
        @Override
        public void onStart() {
            if (mAuthorizationTask != null)
                mExecutorService.submit(mAuthorizationTask);
        }

        @Override
        public void onStop() {
            Dispose.safely(Authorizer.this);
        }
    };

    public Authorizer(MessagingContext messagingContext) {
        mMessagingContext = messagingContext;
        mExecutorService = Executors.newSingleThreadExecutor();
    }

    /**
     * Start authorization process
     * @param identity Identity to use to authorize
     */
    public void start(Identity identity) {
        // Define authorization task.
        // Start will be deferred until the message broker is up.
        mAuthorizationTask = () -> {
            boolean success = false;
            while (true) {
                try {
                    IdentityPublisher isc = new IdentityPublisher(ActiveMQContext.instance());

                    if (identity.hasId()) {
                        // Simply publish id
                        mLog.info(String.format("Publishing [%s]", identity));

                        isc.publish(identity);
                    } else {
                        // Synchronous request for id
                        mLog.info(String.format("Request id for [%s]", identity));
                        AuthorizationMessage authorizationMessage = isc.requestId(identity);

                        // Set id based on response and store identity
                        mLog.info(String.format("Received authorization update [%s]", authorizationMessage));
                        identity.setId(authorizationMessage.getId());
                        identity.store(LocalStorage.instance().getIdentityConfigurationFile());
                    }
                    success = true;
                } catch (TimeoutException e) {
                    mLog.error(e.getMessage());
                } catch (Exception e) {
                    mLog.error(e.getMessage(), e);
                }

                if (success)
                    break;
                else {
                    if (mExecutorService.isShutdown())
                        break;

                    // Retry delay
                    try {
                        TimeUnit.MINUTES.sleep(1);
                    } catch (InterruptedException e1) {
                        mLog.error(e1.getMessage(), e1);
                    }
                }
            }
            mExecutorService.shutdown();
        };

        // Register broker event
        mMessagingContext.getBroker().getDelegate().add(mBrokerEventListener);
        if (mMessagingContext.getBroker().isStarted())
            mExecutorService.submit(mAuthorizationTask);
    }

    @Override
    public void dispose() {
        mExecutorService.shutdownNow();
        try {
            mExecutorService.awaitTermination(Long.MAX_VALUE, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            mLog.error(e.getMessage(), e);
        }
    }
}

File: leoz-node/src/main/java/org/deku/leoz/node/auth/Authorizer.kt
package org.deku.leoz.node.auth

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.commons.logging.LogFactory
import org.deku.leoz.Identity
import org.deku.leoz.config.messaging.MessagingConfiguration
import org.deku.leoz.node.config.StorageConfiguration
import org.deku.leoz.node.messaging.entities.AuthorizationMessage
import org.deku.leoz.node.messaging.entities.AuthorizationRequestMessage
import sx.Disposable
import sx.Dispose
import sx.jms.Channel
import sx.jms.embedded.Broker
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException

/**
 * Authorizer
 * Created by masc on 01.07.15.
 */
class Authorizer(
        /** Messaging context  */
        private val messagingConfiguration: MessagingConfiguration)
:
        Disposable
{
    private val log = LogFactory.getLog(this.javaClass)
    /** Executor service for authorization task  */
    private var executorService: ExecutorService? = null
    /** Authorization task  */
    private var authorizationTask: Runnable? = null

    /** c'or */
    init {
    }

    /** Broker event listener */
    private val brokerEventListener = object : Broker.DefaultEventListener() {
        override fun onStart() {
            if (authorizationTask != null)
                this@Authorizer.executorService!!.submit(authorizationTask)
        }

        override fun onStop() {
            Dispose.safely(this@Authorizer)
        }
    }

    /**
     * Start authorization process.
     * @param identity Identity to use to authorize
     * @param onRejected Rejection callback
     */
    @Synchronized fun start(identity: Identity, onRejected: (identity: Identity) -> Unit = {}) {
        this.stop()
        this.executorService = Executors.newSingleThreadExecutor()

        // Define authorization task.
        // Start will be deferred until the message broker is up.
        authorizationTask = Runnable {
            var success = false
            val executorService = this.executorService!!
            while (!success && !executorService.isShutdown) {
                try {
                    // Setup message
                    val authorizationRequest = AuthorizationRequestMessage()
                    authorizationRequest.name = identity.name
                    authorizationRequest.key = identity.key

                    // Serialize system info to json
                    val jsonMapper = ObjectMapper()
                    val systemInformationJson: String
                    try {
                        systemInformationJson = jsonMapper.writeValueAsString(identity.systemInformation)
                    } catch (e: JsonProcessingException) {
                        throw RuntimeException(e)
                    }

                    authorizationRequest.systemInfo = systemInformationJson

                    log.info("Sending ${authorizationRequest}")

                    // Connection and session
                    val authorizationMessage = Channel(messagingConfiguration.centralQueue).use {
                        it.sendRequest(authorizationRequest).use {
                            it.receive(AuthorizationMessage::class.java)
                        }
                    }

                    // Set id based on response and store identity
                    log.info("Received authorization [%s]".format(authorizationMessage))
                    if (authorizationMessage.rejected) {
                        onRejected(identity)
                    } else {
                        identity.save(StorageConfiguration.instance.identityConfigurationFile)
                    }
                    success = true
                } catch (e: TimeoutException) {
                    log.error(e.message)
                } catch (e: Exception) {
                    log.error(e.message, e)
                }

                if (success)
                    break
                else {
                    if (executorService.isShutdown)
                        break

                    // Retry delay
                    try {
                        TimeUnit.MINUTES.sleep(1)
                    } catch (e1: InterruptedException) {
                        log.error(e1.message, e1)
                    }

                }
            }
            executorService.shutdown()
        }

        // Register broker event
        messagingConfiguration.broker.delegate.add(brokerEventListener)
        if (messagingConfiguration.broker.isStarted)
            executorService!!.submit(authorizationTask)
    }

    /**
     * Stop executor/authorizer and wait for shutdown
     */
    @Synchronized private fun stop() {
        val executorService = this.executorService
        if (executorService != null) {
            executorService.shutdownNow()
            try {
                executorService.awaitTermination(java.lang.Long.MAX_VALUE, TimeUnit.SECONDS)
            } catch (e: InterruptedException) { }
            this.executorService = null
        }
    }

    override fun close() {
        this.stop()
    }
}
